### ISSUE TITLE
Throwing hresults is an unconventional way to return (and the caveats thereof)

### ISSUE THREAD

**Issue by DHowett** [microsoft/cppwinrt#798]
I know that this is a controversial opinion, and that there's no space in the C++/WinRT component authoring interface for making changes here, but I'd be interested in a discussion on this topic.

Terminal recently hooked up a WIL error handling fallback that would report failures to us that weren't appropriately handled elsewhere. That works out really well, except for one thing:

https://github.com/microsoft/terminal/blob/67dfbd21a6b37025ebc0cc435c2ddf791e32811a/src/cascadia/TerminalControl/XamlUiaTextRange.cpp#L99-L104

```c++
            // We _need_ to return XAML_E_NOT_SUPPORTED here.
            // Returning nullptr is an improper implementation of it being unsupported.
            // UIA Clients rely on this HRESULT to signify that the requested attribute is undefined.
            // Anything else will result in the UIA Client refusing to read when navigating by word
            // Magically, this doesn't affect other forms of navigation...
            winrt::throw_hresult(XAML_E_NOT_SUPPORTED);
```

We needed to _simply return an `HRESULT`_ from an implementation type. This is an expected way to implement this particular interface. It works fine, but it's strange because we need to throw an exception to signal that it's okay.

There's three problems with this:

1. Exceptions are not flow control
2. We are originating _thousands_ of errors through `RoOriginateError`. These aren't _really_ errors in any sense of the word, and they do not need to be registered anywhere. They're a signal to the consumer of this interface that something isn't supported.
   * We can't run this under debug w/ `sxe rto` (break on RoOriginate...) because it originates so many errors.
3. Because WIL and cppwinrt cavort in the background to hook up eachothers' exception types, we get _tens of thousands_ of these errors logged to event streams.

So, questions:
* Is there another better way to do this?
* Should we fall back to older tech like WRL for implementing RuntimeClasses?
* Should we implement this one function in a more ABI style (HR return, out parameters) and handle the ABI transitions ourselves?

**Comment by kennykerr** [comment]
Hi Dustin, this is not a new topic but still fun to consider. 

> it's strange because we need to throw an exception to signal that it's _okay_

This is the crux of the problem. The semantics of this interface are poorly designed. Signaling an expected and non-exceptional condition with an error is the problem. WinRT is designed to treat errors as errors and, in most languages, this translates to origination information being collected, propagated, and exceptions being thrown.

> Should we fall back to older tech like WRL for implementing RuntimeClasses?

You could but I don’t know anyone who would consider this an improvement or a more reliable approach. 

> Should we implement this one function in a more ABI style (HR return, out parameters) and handle the ABI transitions ourselves?

Unfortunately, C++/WinRT does not let you spit an interface down the middle (nobody does). You can either implement the entire interface using the projection or using the ABI. So, one approach is to implement the poorly designed interface using the ABI representation and all others using the projection. C++/WinRT lets you freely mix in implementations of COM-style interfaces.

```C++
#include <Windows.Foundation.h>
#include <winrt/Windows.Foundation.h>

struct Thing : winrt::implements<Thing,
	winrt::Windows::Foundation::IStringable,
	ABI::Windows::Foundation::IClosable>
{
	winrt::hstring ToString()
	{
		return L"Thing";
	}

	HRESULT Close()
	{
		return E_FAIL;
	}
};
```

> Is there another better way to do this?

Keep in mind that exceptions themselves are very efficient. As you noted, its treating exceptions as error handling in order to be faithful to WinRT that makes it expensive because of all the error origination that happens along the way. It is still way cheaper than propagating errors via WIL macros but there is a cost. 

For Rust/WinRT I have a different model partly because the Rust language doesn’t have exceptions at all. There is only one way to propagate results and I came up with a model of doing so efficiently by only using the WinRT error origination machinery when propagating a WinRT error. The net effect is that you can return an error code without including that overhead in cases where its not in fact an error that is being communicated. 

We could do something like that with C++/WinRT where a “lite” exception type could be used to sidestep the origination overhead and merely return the HRESULT. It still uses exceptions, but as I have said it is not the exception that is expensive and moving away from exceptions means moving away from modern C++. 

In conclusion, I do believe the interface in question is faulty. It should provide a mechanism for communicating this non-exceptional condition without involving a WinRT error. But if there is enough interest, I’m open to introducing something like this:

```C++
struct Thing : winrt::implements<Thing,
	winrt::Windows::Foundation::IStringable,
	winrt::Windows::Foundation::IClosable>
{
	winrt::hstring ToString()
	{
		return L"Thing";
	}

	void Close()
	{
		throw E_FAIL;
	}
};
```

The raw HRESULT would be translated as not having and not needing any error propagation (no RoOriginateXxx and friends). Alternatively, we could have a special exception type for this.

**Comment by kennykerr** [comment]
Feel free to reopen if there's interest in exploring this further.
